rllib/env/multi_agent_env_runner.py

```diff
@@ -0,0 +1,887 @@
+            # can happen.
+            # Right now we have:
+            #   1. Most times only agents that step get `terminated`, `truncated`
+            #       i.e. the rest we have to check in the episode.
+            #   2. There are edge cases like, some agents terminated, all others
+            #       truncated and vice versa.
+            # See also `MultiAgentEpisode` for handling the `__all__`.
+            if self._all_agents_done(terminateds, truncateds):
+                # Reset all h-states to the model's initial ones b/c we are starting
+                # a new episode.
+                if self.marl_module and self.marl_module.is_stateful():
+                    states = initial_states
+
+                # Finish the episode.
+                for agent_eps in self._episode.agent_episodes.values():
+                    agent_eps.finalize()
+                # TODO (simon): Call here the `MAE.finalize()` method when ready.
+                done_episodes_to_return.append(self._episode)
+                # Create a new episode instance.
+                self._episode = MultiAgentEpisode(agent_ids=self.agent_ids)
+                # Reset the environment.
+                obs, infos = self.env.reset()
+                # Add initial observations and infos.
+                self._episode.add_env_reset(observations=obs, infos=infos)
+                # Reset h-states to the models' initial ones b/c we are starting a new
+                # episode.
+                if self.marl_module:
+                    # TODO (sven, simon): Are there cases where a module overrides its
+                    # own `initial_states` while learning?
+                    states = initial_states
+            else:
+                # Buffer the states for an eventual next `sample()` call.
+                self._states = states
+
+        # Return done episodes ...
+        # TODO (simon): Check, how much memory this attribute uses.
+        self._done_episodes_for_metrics.extend(done_episodes_to_return)
+        # ... and the ongoing episode chunk. Exclude the ongoing episode if
+        # it is only initialized.
+        ongoing_episode: List["MultiAgentEpisode"] = (
+            [self._episode] if self._episode.t > 0 else []
+        )
+        # Also make sure, we return a copy and start new chunks so that callers
+        # of this function do not alter the ongoing and returned episode object.
+        self._episode = self._episode.cut()
+        if ongoing_episode:
+            for agent_eps in ongoing_episode[0].agent_episodes.values():
+                agent_eps.finalize()
+            self._ongoing_episode_for_metrics[ongoing_episode[0].id_].append(
+                ongoing_episode[0]
+            )
+
+        # Record last metrics collection.
+        self._ts_since_last_metrics += ts
+
+        # Return collected episode data.
+        return done_episodes_to_return + ongoing_episode
+
+    def _sample_episodes(
```

Comment:
Note to self: Need to continue reviewing from here. But waiting for author's responses first.